auth/token-managers/iam-assume-token-manager.ts

```diff
@@ -0,0 +1,130 @@
+   * - iamProfileName: (optional) the name of the trusted profile to use (must be specified with iamAccountId)
+   * - iamAccountId: (optional) the ID of the account the trusted profile is in (must be specified with iamProfileName)
+   * - url: (optional) the endpoint URL for the IAM token service (default value: "https://iam.cloud.ibm.com")
+   * - disableSslVerification: (optional) a flag that indicates whether verification of the token server's SSL certificate
+   * should be disabled or not
+   * - headers: (optional) a set of HTTP headers to be sent with each request to the token service
+   * - clientId: (optional) the "clientId" and "clientSecret" fields are used to form a Basic
+   * Authorization header to be included in each request to the token service
+   * - clientSecret: (optional) the "clientId" and "clientSecret" fields are used to form a Basic
+   * Authorization header to be included in each request to the token service
+   * - scope: (optional) the "scope" parameter to use when fetching the bearer token from the token service
+   *
+   * @throws Error: the configuration options are not valid.
+   */
+  constructor(options: Options) {
+    super(options);
+
+    // This just verifies that the API key is provided and is free of common issues.
+    validateInput(options, this.requiredOptions);
+
+    // This validates the assume-specific fields.
+    // Only one of the following three options may be specified.
+    if (!onlyOne(options.iamProfileId, options.iamProfileCrn, options.iamProfileName)) {
+      throw new Error(
+        'Exactly one of `iamProfileName`, `iamProfileCrn`, or `iamProfileId` must be specified.'
+      );
+    }
+
+    // `iamAccountId` may only be specified if `iamProfileName` is also specified.
+    if (Boolean(options.iamProfileName) !== Boolean(options.iamAccountId)) {
+      throw new Error(
+        '`iamProfileName` and `iamAccountId` must be provided together, or not at all.'
+      );
+    }
+
+    // Set class variables from options. If they are 'undefined' in options,
+    // they won't be changed, as they are 'undefined' to begin with.
+    this.iamProfileId = options.iamProfileId;
+    this.iamProfileCrn = options.iamProfileCrn;
+    this.iamProfileName = options.iamProfileName;
+    this.iamAccountId = options.iamAccountId;
+    this.iamDelegate = options.iamDelegate;
+
+    // Create an instance of the IamTokenManager, which will be used to obtain
+    // an IAM access token for use in the "assume" token exchange. Most option
+    // names are shared between these token manager, and extraneous options will
+    // be ignored, so we can pass the options structure to that constructor as-is.
+    this.iamDelegate = new IamTokenManager(options);
+
+    // Set the grant type and user agent for this flavor of authentication.
+    this.formData.grant_type = 'urn:ibm:params:oauth:grant-type:assume';
+    this.userAgent = buildUserAgent('iam-assume-authenticator');
+  }
```

Comment:
Like the authenticator, most of the logic is already taken care of elsewhere, so this was easy to implement. It's similar to the Container authenticator in that way.

Unlike the Container authenticator, I decided not to include any setters/getters for the member variables. We don't include them in the `IamAuthenticator` and I thought consistency with that class was most important. It also didn't seem necessary, but we can add them compatibly later if need be.